### ISSUE TITLE
DeepSEA - Variant effect predictor

### ISSUE THREAD

**Issue by kousikbioinfo** [kipoi/kipoi#310]
Hi, 
  I was wondering why a vcf file is not being recognised by the variant effect predictor model ? if I use bed file, then I am only getting the probability matrix and not e-value and effect (log2 fold change). I would appreciate any help. Thanks.

Cheers,
Kousik

**Comment by kousikbioinfo** [comment]
Thank you very much for your reply. I have tried what you suggested but getting an error mentioned below -

`import vcf
ModuleNotFoundError: No module named 'vcf'`

I though any requited models/libraries should have been installed with `kipoi.postprocessing.variant_effects.snv_predict` , but I am getting this error. Could  you please help me with this ?

**Comment by krrome** [comment]
Did you install kipoi using `pip install kipoi[vep]`? This should install the required packages using pip. If you keep having trouble with that then try running `conda install -c bioconda pyvcf cyvcf2 pybedtools pysam` in your environment which you have set up for DeepSEA. At the depdency installations for variant effect prediction are at the moment not as smooth as they should be, but we are working on fixing - an update should come out within the next few days. For now keep asking and I will try to help.

**Comment by kousikbioinfo** [comment]
Fabulous !! It works now.. Thanks a lot. 
However, I am a bit puzzled to interpret the results, as I am unable to reproduce the results that I am getting from DeepSEA website. In Kipoi output VCF file, I get the following values - 
`
##INFO=<ID=KV:kipoi:DeepSEA/variantEffects:LOGIT_REF
##INFO=<ID=KV:kipoi:DeepSEA/variantEffects:LOGIT_ALT
##INFO=<ID=KV:kipoi:DeepSEA/variantEffects:REF
##INFO=<ID=KV:kipoi:DeepSEA/variantEffects:ALT
##INFO=<ID=KV:kipoi:DeepSEA/variantEffects:LOGIT
##INFO=<ID=KV:kipoi:DeepSEA/variantEffects:DIFF
##INFO=<ID=KV:kipoi:DeepSEA/variantEffects:rID
` 
I am not sure how to interpret the results. In DeepSEA website, I normally get the **Evalue, effect, and probabilities for different alleles.** But I guess I am not getting those values. Do I we need to calculate further to get functional significance score ?

And also the [*.ipynp](https://github.com/kipoi/kipoi/blob/master/nbs/variant_effect_prediction_simple.ipynb) we mentioned above is not working at the moment.

**Comment by kousikbioinfo** [comment]
@krrome I was wondering whether you have any update ! I would appreciate your help. Thank you.

**Comment by krrome** [comment]
Oh I am sorry to keep you waiting, the *effect* score that corresponds to the website is the `deepsea_effect` which can either be set when using the command line `--scores deepsea_effect` or using the python API: `score_variants(..., scores=["deepsea_effect"])`. 
The probabilities for the individual alleles you should get in the  `ref` and `alt`. So all together that would be `--scores ref alt deepsea_effect` using the CLI or `score_variants(..., scores=["ref", "alt", "deepsea_effect"])` using the python API.
I have tested those values for identical outcome, but if you experience inconsistencies please do report that.
We currently can't offer e-values, as they rely on a set of precomputed scores that require a lot of disk space, so we haven't added that in yet I am afraid.

**Comment by kousikbioinfo** [comment]
Thank you very very much for your reply. I will check and update you. Could you please have a look at the page [*.ipynp](https://github.com/kipoi/kipoi/blob/master/nbs/variant_effect_prediction_simple.ipynb), which is not working at the moment? 
Thanks a lot again.

**Comment by krrome** [comment]
Hi @kousikbioinfo,
Thank you very checking so throughly, it was my fault in the last comment on this issue. The output of deepsea_effect is actually infile.vcf.out.logfoldchange * infile.vcf.out.diff , which is the score used for calculating the e-value. The problem with the deepsea paper is that they use three different (but somewhat similar) scores in the same publication and they differ also in when the averaging over forward and reverse-complementation is performed - hence it was complicated to represent all that in a transparent way. 

I have tested the data you have pasted and I could validate the correct replication of the values.  To do the same please use the following script and insert the correct paths for your input files: Please note that I am using the `predict_snvs` function here which should give the same results as the convenience function `score_variants`, but there seems to be a discrepancy between outputs which has escaped all tests - I am currently looking into that.
```python
### Check deepsea against
import kipoi
import pandas as pd
import numpy as np
import copy
import cyvcf2
import os
import pytest

model_root = ""
vcf_path = "input.vcf"
fasta_path = "Homo_sapiens.GRCh37.chromosomes.chr.fa"


def corr2_coeff(A,B):
    # borrowed from: https://stackoverflow.com/questions/30143417/computing-the-correlation-coefficient-between-two-multi-dimensional-arrays/30143754#30143754
    # Rows are expected to be the variables, columns the observations..
    # Rowwise mean of input arrays & subtract from input arrays themeselves
    A_mA = A - A.mean(1)[:,None]
    B_mB = B - B.mean(1)[:,None]
    # Sum of squares across rows
    ssA = (A_mA**2).sum(1)
    ssB = (B_mB**2).sum(1)
    # Finally get corr coeff
    # rows = variables of A
    # columns = variables of B
    return np.dot(A_mA,B_mB.T)/np.sqrt(np.dot(ssA[:,None],ssB[None]))

def test_varaint_effects():
    from kipoi_veff import ModelInfoExtractor, SnvCenteredRg, predict_snvs, score_variants
    from kipoi_veff.scores import DeepSEA_effect
    ####### Now run variant effect prediction like on the web
    model_dir = model_root+"DeepSEA/variantEffects"
    model = kipoi.get_model(model_dir, source="kipoi")
    Dataloader = kipoi.get_dataloader_factory(model_dir, source="kipoi")  

    model_info = ModelInfoExtractor(model, Dataloader)
    vcf_to_region = SnvCenteredRg(model_info)
    dataloader_arguments = {
        "fasta_file": fasta_path,
     }
    res = predict_snvs(model, Dataloader, vcf_path, dataloader_args=dataloader_arguments,
                          batch_size=32,
                          vcf_to_region=vcf_to_region,
                          evaluation_function_kwargs={'diff_types': {'deepsea_effect': DeepSEA_effect("mean")}},
                          return_predictions = True)

    # load results from the pytorch model
    vcf_fh = cyvcf2.VCF(vcf_path, "r")
    indices = []
    for rec in vcf_fh:
        indices.append(rec.ID)


    website_vcf_logfolds = pd.read_csv("infile.vcf.out.logfoldchange", sep=",")
    website_vcf_diff = pd.read_csv("infile.vcf.out.diff", sep=",")

    # Now just align the deepsea results
    website_vcf_logfolds = website_vcf_logfolds.loc[website_vcf_logfolds["name"].isin(indices),:]
    website_vcf_diff = website_vcf_diff.loc[website_vcf_diff["name"].isin(indices),:]
    website_full_pred = website_vcf_logfolds.iloc[:,6:] * website_vcf_diff.iloc[:,6:]

    import pdb
    pdb.set_trace()

    res_pred_here = res['deepsea_effect']
    res_pred_here.index = indices
    res_pred_here = res_pred_here.loc[website_vcf_diff["name"].values, :]

    corrcoeffs = corr2_coeff(res_pred_here.values.T, website_full_pred.values.T) # diagonal should be 1
    assert np.all(np.diag(corrcoeffs)>= 0.99)
```
I have now also updated the ipynbs for variant effect prediction and mutation maps. Thank you very much for thorough testing and please don't  hold back with feedback.

**Comment by kousikbioinfo** [comment]
Thank you very much. I am getting below mentioned error:
`ModuleNotFoundError: No module named 'kipoi_veff'`
I would appreciate, if you could let me know how to install it

**Comment by krrome** [comment]
sure, just make sure you have the latest version of `kipoi` installed via pip in your environment, also when you install the model dependencies with `kipoi env ...` add the `--vep` argument to install the required packages, if also that fails just do `pip install kipoi_veff` in your environment.

**Comment by kousikbioinfo** [comment]
Hi @krrome ,
Sorry for being a little impatient. I have followed your suggestion, but I am getting the below mentioned error. 
```
    if model.postprocessing.get('variant_effects', None) is None:
AttributeError: 'PostProcModelStruct' object has no attribute 'get'
```

**What I have done -**
```
pip install kipoi[vep]
kipoi env create DeepSEA/variantEffects
kipoi env install DeepSEA/variantEffects --vep
```

**Comment by krrome** [comment]
ok I am trying to reproduce your error with your commands, it will take me a moment

**Comment by kousikbioinfo** [comment]
Sure ! Thank you very much.

**Comment by krrome** [comment]
To clarify the meaning of the scores:
`deepsea_effect` is the score that Zhou et al. 2015 used to calculate the e-value - it is therefore not what the website displays as `effect (Log2)` and it is also not the e-value itself. How to use this score:
1) Either use the precomputed values from the DeepSEA website or calculate `deepsea_effect` of 1M randomly selected variants
2) Calculate the `deepsea_effect` of the new variants
3) Calculate the e-value

To get the values that are displayed on the website as `effect (Log2)` the `logit` score of variant effect prediction can be used - since `logit` uses the natural logarithm for calculation the values produced there have to be devided by the constant `np.log(2)` to get the same numerical values that are displayed on the deepsea website.

**Comment by krrome** [comment]
@kousikbioinfo is can we close this issue / is it solved for you?

**Comment by kousikbioinfo** [comment]
@krrome  Yes, you can close it. I think I can reproduce the effect value. Thanks a lot for your help.
